File: app/src/main/java/com/microsoft/msftbreweryworkshop/ui/BreweryListFragment.java
package com.microsoft.msftbreweryworkshop.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.microsoft.msftbreweryworkshop.MSFTBreweryApplication;
import com.microsoft.msftbreweryworkshop.R;
import com.microsoft.msftbreweryworkshop.api.model.BreweryItem;
import com.microsoft.msftbreweryworkshop.api.model.ListResponse;
import com.microsoft.msftbreweryworkshop.model.BreweryListItem;
import com.microsoft.msftbreweryworkshop.service.BreweryService;
import com.microsoft.msftbreweryworkshop.service.ServiceListener;

import java.util.Collections;

public class BreweryListFragment extends Fragment implements ListItemHandler<BreweryListItem>,
    ServiceListener<ListResponse<BreweryItem>> {
    private BreweryListAdapter listAdapter;
    private RecyclerView recyclerView;
    private BreweryService breweryService;

    static BreweryListFragment newInstance() {
        return new BreweryListFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        breweryService = ((MSFTBreweryApplication) getActivity().getApplication()).provideBreweryService();
        listAdapter = new BreweryListAdapter(getContext(), this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_brewery_list, container, false);
        setupRecyclerView(view);
        loadData();
        return view;
    }

    @Override
    public void onSuccess(ListResponse<BreweryItem> response) {
//        listAdapter.items = response.getData();
        listAdapter.notifyDataSetChanged();
    }

    @Override
    public void onFailure(String error) {
        listAdapter.items = Collections.<BreweryListItem>emptyList();
        Toast.makeText(getContext(), error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onItemClick(BreweryListItem item) {
//        DetailActivity.start(getContext(), DetailPageType.BREWERY_DETAIL, item.getId());
    }

    private void loadData() {
        breweryService.getBreweries(this);
    }

    private void setupRecyclerView(View view) {
        recyclerView = view.findViewById(R.id.breweries);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(listAdapter);
    }
}

File: app/src/main/java/com/microsoft/msftbreweryworkshop/ui/BreweryListFragment.kt
package com.microsoft.msftbreweryworkshop.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.microsoft.msftbreweryworkshop.R
import com.microsoft.msftbreweryworkshop.ext.app
import com.microsoft.msftbreweryworkshop.model.BreweryListItem
import com.microsoft.msftbreweryworkshop.model.enums.DetailPageType
import com.microsoft.msftbreweryworkshop.service.BreweryService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import com.microsoft.msftbreweryworkshop.api.model.BreweryItem as ApiBrewery

class BreweryListFragment : Fragment(), ListItemHandler<BreweryListItem> {
    private val listAdapter by lazy { BreweryListAdapter(context!!, this) }

    private lateinit var recyclerView: RecyclerView
    private val breweryService: BreweryService? by lazy { activity?.app?.provideBreweryService() }

    companion object {
        fun newInstance() = BreweryListFragment()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.fragment_brewery_list, container, false)
        setupRecyclerView(view)
        loadData()
        return view
    }

    override fun onItemClick(item: BreweryListItem) {
        context?.let { context ->
            DetailActivity.start(context, DetailPageType.BREWERY_DETAIL, item.id)
        }
    }

    private fun loadData() {
        CoroutineScope(Dispatchers.Default).launch {
            val response = breweryService?.getBreweries()?.await()

            withContext(Dispatchers.Main) {
                listAdapter.items = response ?: emptyList()
                listAdapter.notifyDataSetChanged()
            }
        }
    }

    private fun setupRecyclerView(view: View) {
        recyclerView = view.findViewById<RecyclerView>(R.id.breweries).apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(context)
            adapter = listAdapter
        }
    }
}
